src/inactive-regions.ts

```diff
@@ -0,0 +1,94 @@
+import * as vscode from 'vscode';
+import * as vscodelc from 'vscode-languageclient/node';
+
+import {ClangdContext} from './clangd-context';
+
+// Parameters for the inactive regions (server-side) push notification.
+// Mirrors the structure in the inactive regions proposal for LSP.
+interface InactiveRegionsParams {
+  // The text document that has to be decorated with the inactive regions
+  // information.
+  textDocument: vscodelc.VersionedTextDocumentIdentifier;
+  // An array of inactive regions.
+  regions: vscodelc.Range[];
+}
+
+// Language server push notification providing the inactive regions
+// information for a text document.
+export const NotificationType =
+  new vscodelc.NotificationType<InactiveRegionsParams>(
+    'textDocument/inactiveRegions');
+
+export function activate(context: ClangdContext) {
+  const feature = new InactiveRegionsFeature(context);
+  context.client.registerFeature(feature);
+  context.client.onNotification(NotificationType,
+    feature.handleNotification.bind(feature));
+}
+    
+export class InactiveRegionsFeature implements vscodelc.StaticFeature {
+  private decorationType?: vscode.TextEditorDecorationType;
+  private files: Map<string, vscode.Range[]> = new Map();
+  private context: ClangdContext;
+
+  constructor(context: ClangdContext) {
+    this.context = context;
+  }
+
+  fillClientCapabilities(capabilities: vscodelc.ClientCapabilities) {
+    // Extend the ClientCapabilities type and add inactive regions
+    // capability to the object.
+    if (capabilities.textDocument) {
+      const textDocumentCapabilities: vscodelc.TextDocumentClientCapabilities &
+      { inactiveRegionsCapabilities?: { inactiveRegions: boolean } } =
+        capabilities.textDocument;
+      textDocumentCapabilities.inactiveRegionsCapabilities = {
+        inactiveRegions: true,
+      };
+    }
+  }
+  initialize(capabilities: vscodelc.ServerCapabilities,
+    documentSelector: vscodelc.DocumentSelector | undefined) {
+    const serverCapabilities: vscodelc.ServerCapabilities & {
+      inactiveRegionsProvider?: any
+    } = capabilities;
+    if (serverCapabilities.inactiveRegionsProvider) {
+      this.decorationType = vscode.window.createTextEditorDecorationType({
+        isWholeLine: true,
+        light: {
+          backgroundColor: "rgba(220, 220, 220, 0.3)",
```

Comment:
A native question, how do you come up with these numbers? I guess we want to follow the inactive-style of the TS in vscode, but I didn't manage to figure out the color style there (the `Developer: Inspect Editor Tokens and Scopes`
command doesn't seem to give useful information there) .

Reply:
The colors also date back to the [2019 patch](https://reviews.llvm.org/D67537?vs=on&id=220043#toc). My best guess as to where I got them from is the colors used by Eclipse CDT in this situation.

That said, your comment reminds me that we should probably make these colors customizable via vscode settings.

Another customization that we could allow is to optionally use a reduced opacity of the foreground colors, instead of a background color (some users have expressed a preference for this visual style).

Reply:
VSCode already provides `editorUnnecessaryCode.border` and `editorUnnecessaryCode.opacity` fields (see https://github.com/microsoft/vscode/blob/main/src/vs/editor/common/core/editorColorRegistry.ts#L44-L45), they are used for unused diagnostics, the rendering style looks very similar to what we propose here (except the background color) and they are theme-aware, can we just use them in our extension to render the inactive regions rather than adding our own ones?  

The downside is that we don't provide many customizations (no foreground&background color), it is fine to not supporting that from my perspective.

Reply:
I am open to using these instead of our own opacity setting in the `useOpacity=true` branch. (I would prefer to keep the `useOpacity=false` branch with the background highlight as an option for users who prefer this look.)

However, I'm not quite sure how to make use of this from the extension. I can create a `new vscode.ThemeColor('editorUnnecessaryCode.opacity')`, but what do I do with it?
 * The `ThemeColor` does not expose the actual color or its opacity
 * The `opacity` field of `ThemableDecorationRenderOptions` does not accept a `ThemeColor`.
 * The `color` field of `ThemableDecorationRenderOptions` accepts a `ThemeColor`, but if I pass `new vscode.ThemeColor('editorUnnecessaryCode.opacity')` to it, then it uses this as the foreground color for inactive regions. While the color does have a reduced opacity, it also overwrites the colors from the textmate tokens, thus defeating the purpose of this patch (which is to combine the opacity with the colors of the textmate tokens).

Is there some other way of using this opacity field that I'm overlooking?

Reply:
Thanks for exploring it. I thought these fields are exposed and can be accessed via the VSCode public API, looks like they are not.